src/middleware.ts

```diff
@@ -605,6 +614,28 @@ export const persist =
       onHydrate: (cb) => {
         onHydrateCallbacks.push(cb)
       },
+      onFinishHydration: (cb) => {
+        onFinishHydrationCallbacks.push(cb)
+      },
+      useHydration: () => {
```

Comment:
I don't think it's a good idea for `zustand/middleware` to depend on `react`. All middlewares including persist should be able to be used in vanilla.

So, I would suggest to not have it in persist api, but in docs as a reference implementation.
Also, if we use a hook, useEffect should be able to unsubscribe the callbacks.

```js
const useStore = create(persist(...))
const useHydration = () => {
  const [hydrated, setHydrated] = useState(useStore.persist.hasHydrated)
  useEffect(() => {
    const unsub1 = useStore.persist.onHydrate(() => setHydrated(false)) // not sure this is required in practice
    const unsub2 = useStore.persist.onFinishHydration(() => setHydrated(true))
    return () => {
      unsub1()
      unsub2()
    }
  }
  return hydrated
}
```

Reply:
another option is to provide an api so that people can create a hook more easily.
```js
const useHydration = () => {
  const [hydrated, setHydrated] = useState(useStore.persist.hasHydrated)
  useEffect(() => useStore.persist.onChangeHydrated(setHydrated), [])
  return hydrated
}
```

Reply:
On second thought, why would we need another hook while we already have a store hook.
So, I'd expect this is better practice.
```js
const useStore = create(persist(...))
useStore.persist.onFinishHydration(() => useStore.setState({ hydrated: true }))
```
This may not work if the hydration is sync? Maybe better to have a persist option, instead of using persist api?

Hm, we already have postRehydrationCallback. So:
```js
const useStore = create(persist(
  (set) => ...,
  {
    name: 'name',
    onRehydrateStorage: () => (state) => {
      if (state) { useStore.setState({ hydrated: true }) }
    },
  }
))
```
It looks tricky, though. Oh, and we need to exclude it when serializing... 😕

Reply:
Surely that does make sens. I wasn't too sure about this and was looking forward to your feedbacks :smile: 

I think we should go with your first example (fully custom `useHydration` hook referenced in the doc). I'll remove the hook and add an unsub function.

> ```useStore.persist.onHydrate(() => setHydrated(false)) // not sure this is required in practice```

FYI this is just in case you want to take into account manual hydrations using `useStore.persist.rehydrate()`

Reply:
Sounds good.